File: ui-common/src/main/java/com/facebook/fresco/ui/common/ControllerListener2.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.ui.common;

import android.net.Uri;
import com.facebook.infer.annotation.Nullsafe;
import com.facebook.infer.annotation.PropagatesNullable;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import javax.annotation.Nullable;

@Nullsafe(Nullsafe.Mode.STRICT)
public interface ControllerListener2<INFO> {

  class Extras {
    public @Nullable Map<String, Object> componentExtras;
    public @Nullable Map<String, Object> shortcutExtras;
    public @Nullable Map<String, Object> datasourceExtras;
    public @Nullable Map<String, Object> imageExtras;

    public @Nullable Object callerContext;
    public @Nullable Uri mainUri;

    public int viewportWidth = -1;
    public int viewportHeight = -1;
    public @Nullable Object scaleType;
    public float focusX = -1;
    public float focusY = -1;

    public static Extras of(@Nullable Map<String, Object> componentExtras) {
      Extras extras = new Extras();
      extras.componentExtras = componentExtras;
      return extras;
    }

    public Extras makeExtrasCopy() {
      Extras extras = new Extras();
      extras.componentExtras = copyMap(this.componentExtras);
      extras.shortcutExtras = copyMap(this.shortcutExtras);
      extras.datasourceExtras = copyMap(this.datasourceExtras);
      extras.imageExtras = copyMap(this.imageExtras);
      extras.callerContext = this.callerContext;
      extras.mainUri = this.mainUri;
      extras.viewportWidth = this.viewportWidth;
      extras.viewportHeight = this.viewportHeight;
      extras.scaleType = this.scaleType;
      extras.focusX = this.focusX;
      extras.focusY = this.focusY;

      return extras;
    }

    private static Map<String, Object> copyMap(@PropagatesNullable Map<String, Object> map) {
      if (map == null) {
        return null;
      }

      return new ConcurrentHashMap<>(map);
    }
  }

  /**
   * Called before the image request is submitted.
   *
   * <p>IMPORTANT: It is not safe to reuse the controller from within this callback!
   *
   * @param id controller id
   * @param callerContext caller context
   * @param extraData extra data
   */
  void onSubmit(String id, Object callerContext, @Nullable Extras extraData);

  /**
   * Called after the final image has been set.
   *
   * @param id controller id
   * @param imageInfo image info
   * @param extraData extra data
   */
  void onFinalImageSet(String id, @Nullable INFO imageInfo, Extras extraData);

  /**
   * Called after any intermediate image has been set.
   *
   * @param id controller id
   */
  void onIntermediateImageSet(String id, @Nullable INFO imageInfo);

  /**
   * Called after the fetch of the intermediate image failed.
   *
   * @param id controller id
   */
  void onIntermediateImageFailed(String id);

  /**
   * Called after the fetch of the final image failed.
   *
   * @param id controller id
   * @param throwable failure cause
   * @param extraData extra data
   */
  void onFailure(String id, Throwable throwable, @Nullable Extras extraData);

  /**
   * Called after the controller released the fetched image.
   *
   * <p>IMPORTANT: It is not safe to reuse the controller from within this callback!
   *
   * @param id controller id
   * @param extraData extra data
   */
  void onRelease(String id, @Nullable Extras extraData);
}

File: ui-common/src/main/java/com/facebook/fresco/ui/common/ControllerListener2.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.ui.common

import android.net.Uri
import com.facebook.infer.annotation.PropagatesNullable
import java.util.concurrent.ConcurrentHashMap
import kotlin.jvm.JvmField

interface ControllerListener2<INFO> {

  class Extras {
    @JvmField var componentExtras: Map<String, Any>? = null
    @JvmField var shortcutExtras: Map<String, Any>? = null
    @JvmField var datasourceExtras: Map<String, Any>? = null
    @JvmField var imageExtras: Map<String, Any>? = null
    @JvmField var imageSourceExtras: Map<String, Any>? = null
    @JvmField var callerContext: Any? = null
    @JvmField var mainUri: Uri? = null
    @JvmField var viewportWidth: Int = -1
    @JvmField var viewportHeight: Int = -1
    @JvmField var scaleType: Any? = null
    @JvmField var focusX: Float? = null
    @JvmField var focusY: Float? = null
    @JvmField var logWithHighSamplingRate: Boolean = false
    @JvmField var modifiedUriStatus: String? = null
    @JvmField var originalUri: Uri? = null
    @JvmField var uiFramework: String? = null

    fun makeExtrasCopy(): Extras {
      val extras = Extras()
      extras.componentExtras = copyMap(componentExtras)
      extras.shortcutExtras = copyMap(shortcutExtras)
      extras.datasourceExtras = copyMap(datasourceExtras)
      extras.imageExtras = copyMap(imageExtras)
      extras.callerContext = callerContext
      extras.mainUri = mainUri
      extras.viewportWidth = viewportWidth
      extras.viewportHeight = viewportHeight
      extras.scaleType = scaleType
      extras.focusX = focusX
      extras.focusY = focusY
      extras.uiFramework = uiFramework
      return extras
    }

    companion object {
      @JvmStatic
      fun of(componentExtras: Map<String, Any>?): Extras {
        val extras = Extras()
        extras.componentExtras = componentExtras
        return extras
      }

      private fun copyMap(@PropagatesNullable map: Map<String, Any>?): Map<String, Any>? =
          map?.let(::ConcurrentHashMap)
    }
  }

  /**
   * Called before the image request is submitted.
   *
   * IMPORTANT: It is not safe to reuse the controller from within this callback!
   *
   * @param id controller id
   * @param callerContext caller context
   * @param extraData extra data
   */
  fun onSubmit(id: String, callerContext: Any?, extraData: Extras?)

  /**
   * Called after the final image has been set.
   *
   * @param id controller id
   * @param imageInfo image info
   * @param extraData extra data
   */
  fun onFinalImageSet(id: String, imageInfo: INFO?, extraData: Extras?)

  /**
   * Called after any intermediate image has been set.
   *
   * @param id controller id
   */
  fun onIntermediateImageSet(id: String, imageInfo: INFO?)

  /**
   * Called after the fetch of the intermediate image failed.
   *
   * @param id controller id
   */
  fun onIntermediateImageFailed(id: String)

  /**
   * Called after the fetch of the final image failed.
   *
   * @param id controller id
   * @param throwable failure cause
   * @param extraData extra data
   */
  fun onFailure(id: String, throwable: Throwable?, extraData: Extras?)

  /**
   * Called after the controller released the fetched image.
   *
   * IMPORTANT: It is not safe to reuse the controller from within this callback!
   *
   * @param id controller id
   * @param extraData extra data
   */
  fun onRelease(id: String, extraData: Extras?)

  fun onEmptyEvent(callerContext: Any?)
}
